internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,311 @@
+	// are added as attributes to this address. The resulting list of addresses
+	// is then grouped into endpoints, covering all combinations of proxy and
+	// target endpoints.
+	var endpoints []resolver.Endpoint
+	for _, proxyEndpt := range r.proxyEndpoint {
+		proxyAddr := proxyEndpt.Addresses[0].Addr
+		for _, endpt := range r.targetEndpoint {
+			var addrs []resolver.Address
+			for _, targetAddr := range endpt.Addresses {
+				newAddr := resolver.Address{Addr: proxyAddr}
+				newAddr = proxyattributes.Populate(newAddr, proxyattributes.Options{
+					User:        r.proxyURL.User,
+					ConnectAddr: targetAddr.Addr,
+				})
+				addrs = append(addrs, newAddr)
+			}
+			endpoints = append(endpoints, resolver.Endpoint{Addresses: addrs})
+		}
+	}
+	return addresses, endpoints
+}
+
+// updateProxyResolverState updates the proxy resolver state by storing proxy
+// addresses and endpoints, marking the resolver as ready, and triggering a
+// state update if both proxy and target resolvers are ready. It is a
+// StateListener function of wrappingClientConn passed to the proxy resolver.
+func (r *delegatingResolver) updateProxyResolverState(state resolver.State) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if logger.V(2) {
+		logger.Infof("Addresses received from proxy resolver: %s", state.Addresses)
+	}
+	r.proxyAddrs = state.Addresses
+	r.proxyEndpoint = state.Endpoints
+	r.proxyResolverReady = true
+	r.updateStateIfReadyLocked()
+	return nil
+}
+
+// updateTargetResolverState updates the target resolver state by storing target
+// addresses, endpoints, and service config, marking the resolver as ready, and
+// triggering a state update if both resolvers are ready. It is a StateListener
+// function of wrappingClientConn passed to the target resolver.
+func (r *delegatingResolver) updateTargetResolverState(state resolver.State) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if logger.V(2) {
+		logger.Infof("Addresses received from target resolver: %v", state.Addresses)
+	}
+	r.targetAddrs = state.Addresses
+	r.targetEndpoint = state.Endpoints
+	r.targetResolverReady = true
+	// Update curState to include other state information, such as the
+	// service config, provided by the target resolver. This ensures
+	// curState contains all necessary information when passed to
+	// UpdateState. The state update is only sent after both the target and
+	// proxy resolvers have sent their updates, and curState has been
+	// updated with the combined addresses.
+	r.curState = state
+	r.updateStateIfReadyLocked()
+	return nil
+}
+
+// updateStateIfReadyLocked checks if both proxy and target resolvers are ready,
+// combines addresses, and updates the ClientConn state.
+func (r *delegatingResolver) updateStateIfReadyLocked() {
+	if !r.targetResolverReady || !r.proxyResolverReady {
+		return
+	}
+	r.curState.Addresses, r.curState.Endpoints = r.combinedAddressesLocked()
+	r.cc.UpdateState(r.curState)
```

Comment:
We should not discard the error returned by the channel, it is used to make the resolver backoff.

Reply:
+1

Reply:
Done.